File: Code/CSharp/Tools/Rorn.Tools.Maths/BoundingBox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Rorn.Tools.Maths
{
    public class BoundingBox
    {
        public BoundingBox()
        {
            MinX = 0;
            MinY = 0;
            MinZ = 0;
            MaxX = 0;
            MaxY = 0;
            MaxZ = 0;
            initialized_ = false;
        }

        public void AddPoint(Vector4 point)
        {
            if (initialized_)
            {
                if (point.X < MinX)
                    MinX = point.X;

                if (point.Y < MinY)
                    MinY = point.Y;

                if (point.Z < MinZ)
                    MinZ = point.Z;

                if (point.X > MaxX)
                    MaxX = point.X;

                if (point.Y > MaxY)
                    MaxY = point.Y;

                if (point.Z > MaxZ)
                    MaxZ = point.Z;
            }
            else
            {
                // This is the first point
                MinX = point.X;
                MinY = point.Y;
                MinZ = point.Z;
                MaxX = point.X;
                MaxY = point.Y;
                MaxZ = point.Z;
                initialized_ = true;
            }
        }

        public void Save(System.IO.BinaryWriter binaryWriter)
        {
            binaryWriter.Write(MinX);
            binaryWriter.Write(MinY);
            binaryWriter.Write(MinZ);
            binaryWriter.Write(MaxX);
            binaryWriter.Write(MaxY);
            binaryWriter.Write(MaxZ);
        }

        public float MinX { get; private set; }
        public float MinY { get; private set; }
        public float MinZ { get; private set; }
        public float MaxX { get; private set; }
        public float MaxY { get; private set; }
        public float MaxZ { get; private set; }

        private bool initialized_;
    }
}

File: Code/CPlusPlus/Shared/RornMaths/BoundingBox.cpp
#include "BoundingBox.h"

#include <algorithm>

using namespace Rorn::Maths;

BoundingBox::BoundingBox()
	: minimum_(0.0f, 0.0f, 0.0f, 1.0f), maximum_(0.0f, 0.0f, 0.0f, 1.0f)
{
}

BoundingBox::BoundingBox(float minX, float minY, float minZ, float maxX, float maxY, float maxZ)
	: minimum_(minX, minY, minZ, 1.0f), maximum_(maxX, maxY, maxZ, 1.0f)
{
}

float BoundingBox::GetBoundingRadius() const
{
	float sizeX = maximum_.X - minimum_.X;
	float sizeY = maximum_.Y - minimum_.Y;
	float sizeZ = maximum_.Z - minimum_.Z;

	// just return half the biggest dimension
	return (std::max(sizeX, std::max(sizeY, sizeZ) ) * 0.5f);
}

Vector4 BoundingBox::GetMinimum() const
{
	return minimum_;
}

Vector4 BoundingBox::GetMaximum() const
{
	return maximum_;
}

Vector4 BoundingBox::GetCentre() const
{
	return Vector4(
		minimum_.X + ((maximum_.X - minimum_.X) / 2.0f),
		minimum_.Y + ((maximum_.Y - minimum_.Y) / 2.0f),
		minimum_.Z + ((maximum_.Z - minimum_.Z) / 2.0f),
		1.0f);
}
